### PR TITLE
Add https request error handling

### PR DESCRIPTION
**Pull request by cjh980402**
When I did some test, https request error handling is necessary.

### CODE REVIEW

**Comment by killer069** [comment]
> When I did some test, https request error handling is necessary.

Can you explain why ?

**Comment by cjh980402** [comment]
If is there are no `https request error handling`, then some https request error or network error is occured by unhandled promise rejection.

**Comment by killer069** [comment]
> If is there are no `https request error handling`, then some https request error or network error is occured by unhandled promise rejection.

If you are using discord.js, then we can confirm that the bot has internet access and also a good internet. No need to handle that.

**Comment by cjh980402** [comment]
I think it is right to add error handling because this module can be used independently other than discord.js.
Because play-dl is good module to download youtube.

**Comment by killer069** [comment]
> I think it is right to add error handling because this module can be used independently other than discord.js.
> Because play-dl is good module to download youtube.

I don't recommend this as a module to download youtube since I have added that pause system which will only let users to download only first 5 mins part only.

**Comment by cjh980402** [comment]
Um...okay. Then, how about other code (remove useless code, wrong formatting and some modification of option parameter)

File: play-dl/YouTube/utils/request.ts [hunk 1]
```diff
@@ -10,45 +10,55 @@ interface RequestOpts extends RequestOptions{
...
     })
 }
 
 export async function request(url : string, options? : RequestOpts): Promise<string>{
     return new Promise(async (resolve, reject) => {
-        let data = ''
-        let res = await https_getter(url, options)
-        if(Number(res.statusCode) >= 300 && Number(res.statusCode) < 400){
-            res = await https_getter(res.headers.location as string , options)
+        try{
```
Comment (killer069) [inline play-dl/YouTube/utils/request.ts:32]: Remove try and catch error.

File: play-dl/YouTube/utils/request.ts [hunk 2]
```diff
@@ -10,45 +10,55 @@ interface RequestOpts extends RequestOptions{
...
     })
 }
 
 export async function request_stream(url : string, options? : RequestOpts): Promise<IncomingMessage>{
     return new Promise(async (resolve, reject) => {
-        let data = ''
-        let res = await https_getter(url, options)
-        if(Number(res.statusCode) >= 300 && Number(res.statusCode) < 400){
-            res = await https_getter(res.headers.location as string, options)
+        try{
```
Comment (killer069) [inline play-dl/YouTube/utils/request.ts:53]: Here also

**Review by killer069 — CHANGES_REQUESTED** [review]
Do the changes and then I will merge.

**Review by killer069 — APPROVED** [review]

**Comment by cjh980402** [comment]
```js
private async loop(){
        if(this.stream.destroyed){
            this.cleanup()
            return
        }
        let end : number = this.bytes_count + this.per_sec_bytes * 300;
        let stream = await request_stream(this.url, {
            headers : {
                "range" : `bytes=${this.bytes_count}-${end >= this.content_length ? '' : end}`
            }
        })
        if(Number(stream.statusCode) >= 400){
            this.cleanup()
            await this.retry()
            this.loop()
        }
        this.request = stream
        stream.pipe(this.stream, { end : false })

        stream.once('error', (err) => {
            this.stream.emit('error', err)
        })

        stream.on('data', (chunk: any) => {
            this.bytes_count += chunk.length
        })

        stream.on('end', () => {
            if(end >= this.content_length) this.data_ended = true
        })
    }
```
this is loop of Stream, and I have some question.
```js
if(Number(stream.statusCode) >= 400){
            this.cleanup()
            await this.retry()
            this.loop()
        }
```
this part can makes something wrong coz cleanup makes `this.bytes_count` and `this.per_sec_bytes` to zero and there are no function break.

**Comment by cjh980402** [comment]
```js
export class Stream {
    type : StreamType
    stream : PassThrough
    private url : string
    private bytes_count : number;
    private per_sec_bytes : number;
    private content_length : number;
    private video_url : string;
    private timer : NodeJS.Timer;
    private cookie : string;
    private data_ended : boolean;
    private playing_count : number;
    private request : IncomingMessage | null
    constructor(url : string, type : StreamType, duration : number, contentLength : number, video_url : string, cookie : string){
        this.url = url
        this.type = type
        this.stream = new PassThrough({ highWaterMark : 10 * 1000 * 1000 })
        this.bytes_count = 0
        this.video_url = video_url
        this.cookie = cookie
        this.timer = setTimeout(() => {
            this.retry()
        }, 7200 * 1000)
        this.per_sec_bytes = Math.ceil(contentLength / duration)
        this.content_length = contentLength
        this.request = null
        this.data_ended = false
        this.playing_count = 0
        this.stream.on('close', () => {
            this.cleanup()
        })
        this.stream.on('pause', () => {
            this.playing_count++;
            if(this.data_ended){
                this.cleanup()
                this.stream.removeAllListeners('pause')
            }
            else if(this.playing_count === 280){
                this.playing_count = 0
                this.loop()
            }
        })
        this.loop()
    }

    private async retry(){
        let info = await video_info(this.video_url, this.cookie)
        this.url = info.format[info.format.length - 1].url
        this.timer = setTimeout(() => {
            this.retry()
        }, 7200 * 1000)
    }

    private cleanup(){
        clearInterval(this.timer)
        this.request?.unpipe(this.stream)
        this.request?.destroy()
        this.request = null
        this.url = ''
    }

    private async loop(){
        if(this.stream.destroyed){
            this.cleanup()
            return
        }
        let end : number = this.bytes_count + this.per_sec_bytes * 300;
        let stream = await request_stream(this.url, {
            headers : {
                "range" : `bytes=${this.bytes_count}-${end >= this.content_length ? '' : end}`
            }
        })
        if(Number(stream.statusCode) >= 400){
            this.cleanup()
            await this.retry()
            this.loop()
            return
        }
        this.request = stream
        stream.pipe(this.stream, { end : false })

        stream.once('error', (err) => {
            this.stream.emit('error', err)
        })

        stream.on('data', (chunk: any) => {
            this.bytes_count += chunk.length
        })

        stream.on('end', () => {
            if(end >= this.content_length) this.data_ended = true
        })
    }
}
```
How about this? Change setInterval to setTimeout, retry can every stream, function break to do not run pipe and other events.

**Comment by killer069** [comment]
> How about this? Change setInterval to setTimeout to retry can every stream, function break to do not run pipe and other events.

This would start creating a new loop if you hit 404 errors too.

**Comment by killer069** [comment]
> How about this? Change setInterval to setTimeout to retry can every stream, function break to do not run pipe and other events.

Good Idea. 👍

**Pull request merged**